Function to determine the number of frames

ezavod opened the pull request:
This is needed to replace molfiles with this library, because the number of frames needs to be known for `Trajectory::nsteps()`. This isn't really elegant, but it is as good as the current solution via molfiles.
Future improvements: use a faster seek method to find all frames.

ezavod committed:
feat(TRR/XTC): read header info

This adds functions to read the initial number of atoms and number of
frames.

ezavod committed:
fix(TRR): correct return code and magic check

See `xtc_header(...)` in `xdrfile_xtc.c`.

Luthaf commented on src/xdrfile_trr.c:
```diff
@@ -37,7 +37,7 @@
 
 #define BUFSIZE 128
-#define GROMACS_MAGIC 1993
+#define MAGIC 1993
```
Why the rename? It is still a GROMACS format :smile:

ezavod replied:
I was looking at the xtc and trr code side by side. The xtc magic number is simply called `MAGIC` so I changed it for easier readability. Didn't think about that much. Should it be reverted?

Luthaf replied:
oh, ok =) Maybe naming them `XTC_MAGIC` and `TRR_MAGIC` is the best =)

ezavod replied:
Good idea :thumbsup:

Luthaf reviewed (commented):
I don't think these functions belong to the xdrfile library: we could like in chemfiles to record the file offset of the beginning of each frame; meaning we would need to write the same loop in chemfiles, to get both the number of frames and the file offsets.

I think we would instead want some form of xdr_tell and xdr_seek function instead.

ezavod commented:
I agree that getting the frame offsets is the main improvement we should go for, because this would speed up jumps between frames. AFAIK this lib has `xdrstdio_getpos` and `xdrstdio_setpos`. Those could be used to get the offsets and seek I guess. This could be done in `read_*_header` since this already loops over all the frames (as you said duplicating this would be bad).
So in the end `read_*_header` would return number of atoms, frames and corresponding offsets. I didn't had the time to implement the offsets right away.

I see multiple reasons, why this `read_*_header` function should be part of xdrfile:

- `read_*_natoms` is already part of xdrfile and also is some sort of helper function. `read_*_header` is just another helper function.
- other forks of the GROMACS version have similar functions at the same place e.g. cpptraj (thats admittedly a weak argument).
- this is a general improvement to the library itself. If any one else wants to use this library in their projects, this could be helpful.
- putting the offset calculation in chemfiles would add unnecessary complexity over there. I would argue, that getting the number of frames and offsets is an integral part of a xtc/trr library. This shouldn't be the task of an other project that simply wants to read the trajectory.

IMHO the last two are the important ones.
This seems to be a question of philosophy. Feel free to do what ever fits the structure of your project best :)

I could also look into the offset calculation and add it to this PR. But this might take some time unfortunately.

Luthaf commented:
A function that reads both the number of frame and the frame offsets would effectively be more useful 😃 

> AFAIK this lib has xdrstdio_getpos and xdrstdio_setpos. Those could be used to get the offsets and seek I guess

Unfortunately we can't use these functions, since they use 32-bit offset, as as such are limited to ~2Gb files. 

I spent the day playing with this library (and #3 resulted), and I have an implementation of 64-bit offsets seeking that I'll send soon.

A nice trick to read frames offsets (and thus the number of frame) is to only read the frame header, and compute the size of (possibly compressed) data that needs to be skipped to get to the next frame, instead of decoding and reading the full frame.

ezavod commented:
> Unfortunately we can't use these functions, since they use 32-bit offset, as as such are limited to ~2Gb files.

Uh, missed that. That wouldn't be good. Looking forward for your 64bit version then.

>A nice trick to read frames offsets (and thus the number of frame) is to only read the frame header...

That would be great indeed, but it's not apparent how long the compressed data is. Will play around with that a bit.

ezavod committed:
feat(XDR): seek and tell with 64bit positions

ezavod committed:
feat(TRR/XTC): basic offset calculation by parsing every frame

Iterates over all frames and stores the corresponding offsets for later
use. Enables fast reading of a specific step.

ezavod committed:
feat(TRR): skip frames for offsets

Calculate the size of the frame to skip the contents. Only needs to
parse every header instead of the whole frame. Also estimates the number
of frames in order to minimize allocations.

ezavod committed:
feat(XTC): skip frames for offsets

Calculate the size of the frame to skip the contents. Only needs to
parse every framesize instead of the whole frame. Also estimates the number
of frames in order to minimize allocations.

ezavod commented:
This should be all that's needed for full offset support :tada:

To make future life easier some references:
- seek/tell compile flags: [MDAnalysis](https://github.com/MDAnalysis/mdanalysis/blob/78540119b3137f8407a31a95dd1cca0c79ceef19/package/setup.py#L285), [Doc](http://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#Feature-Test-Macros)
- on rounding to 32bit boundary: [explanation on SO](https://stackoverflow.com/questions/8178140/way-to-round-up-number-to-an-int-size-boundary-number-of-bytes), there seems to be a problem for systems that are not based on power of 2 datatype sizes, but this seems to be super rare ([alternative](https://stackoverflow.com/questions/3407012/c-rounding-up-to-the-nearest-multiple-of-a-number/9194117#9194117)).

ezavod committed:
refactor(TRR/XTC): rename magic constant

Luthaf commented on src/xdrfile.c:
```diff
@@ -2120,11 +2125,41 @@ static int xdrstdio_putbytes(XDR* xdrs, char* addr, unsigned int len) {
-    return (unsigned int)ftell((FILE*)xdrs->x_private);
+/* 64 bit fileseek operations */
+static int64_t xdrstdio_getpos(XDR* xdrs) {
+#ifdef _WIN32
```
We may want to add an ifdef for cygwin (see https://stackoverflow.com/a/30867832), as it does not  offer ftello. At the same time, I would prefer using `ifdef _MSC_VER` instead of `_WIN32`, since we can have non-MSVC compilers on windows which can define _WIN32.

ezavod replied:
This makes sense.
Some people seem to have some issues with `_ftelli64` and `fopen` (which we are using). Unfortunately I have no way to test this. If that's only a problem if one seeks from the end ([SO](https://stackoverflow.com/questions/4034227/on-windows-fseeki64-does-not-seek-to-seek-end-correctly-for-large-files)), we should be fine (would only lead to wrong frame number estimation).

Luthaf replied:
Can we use `GetFileSizeEx()` instead on windows to get the file size, as suggested on SO?

Luthaf replied:
And continue using _ftelli64/fopen for the rest of the library?

ezavod replied:
We could use `GetFileSizeEx()` for getting the file size, but if `_ftelli64` is broken, there are the still uses for getting the frame offset. Even an error of a few bytes (as reported in your SO) would break this. _If_ it's still broken, then we have to find a way to replace `_ftelli64/_fseeki64` altogether.
The posts are relatively old (4 and 9 years), so maybe we are lucky and its fixed =)

Luthaf replied:
ok, I'll try to run the tests on Windows to check if this works now

Luthaf commented on src/xdrfile.c:
```diff
@@ -2120,11 +2125,41 @@ static int xdrstdio_putbytes(XDR* xdrs, char* addr, unsigned int len) {
+/* We return errno relying on the fact that it is never set to 0 on
+ * success, which means that if an error occurrs it'll never be the same
+ * as exdrOK, and xdr_seek won't be confused.*/
+#ifdef _WIN32
```
Same comment as `xdrstdio_getpos`

Luthaf commented on include/xdrfile_trr.h:
```diff
@@ -38,18 +38,40 @@ extern "C" {
+/* This function returns the number of frames and the number of atoms
+ * in the trr file in *natoms and *nframes
+ */
+int read_trr_header(const char* fn, int* natoms, unsigned long* nframes, int64_t** offsets);
```
Could you add the description of `offsets` here, as well as the fact that users are responsible to free the associated memory?

ezavod replied:
Done.

Luthaf commented on include/xdrfile_xtc.h:
```diff
@@ -42,6 +42,11 @@ extern "C" {
 extern int read_xtc_natoms(const char* fn, int* natoms);
 
+/* This function returns the number of frames and the number of atoms
+ * in the xtc file in *natoms and *nframes
```
Same as read_trr_header =)

ezavod replied:
Done.

Luthaf reviewed (changes requested):
The code looks good to me, could you add some tests that the new funtions are working as intended?

Luthaf commented:
> on rounding to 32bit boundary: explanation on SO, there seems to be a problem for systems that are not based on power of 2 datatype sizes, but this seems to be super rare (alternative).

Can this be checked somehow at compile time, for example using #if (...) #error "wrong size" #endif?

ezavod commented:
> Can this be checked somehow at compile time, for example using #if (...) #error "wrong size" #endif?

I don't know any easy solution. [limits.h](http://www.cplusplus.com/reference/climits/) defines `INT_MAX`, so one could test if this suits to a power of 2 (not sure if possible as compiler directive). Or we could just check if this is a 16, 32, 64bit value so we would cover the majority of systems (see [Word Size](https://en.wikipedia.org/wiki/Word_(computer_architecture)#Size_families)).
IMHO this is a bit too much effort for a pretty much non existing edge case. Cant find any of such systems that are newer than the 70s :D

Luthaf commented:
> IMHO this is a bit too much effort for a pretty much non existing edge case. Cant find any of such systems that are newer than the 70s :D

Fair enough =). Maybe we can add a check in chemfiles' CMakeLists.txt that the system is either 32 or 64-bit, since that's all that is supported.

ezavod committed:
feat(TRR/XTC): added tests for offsets

ezavod commented:
[CMAKE_SIZEOF_VOID_P](https://cmake.org/cmake/help/v3.0/variable/CMAKE_SIZEOF_VOID_P.html) seems to be used to determine if one is on a 64 bit system.
But I think that it's possible to have a different size of `int` (which is what matters in our case) so [CheckTypeSize](https://cmake.org/cmake/help/latest/module/CheckTypeSize.html) would be a better fit. I don't think that it's a good thing to limit this lib to systems with a specific size of `int`. Ideally we would be able to test if it's a power of 2 instead of limiting to a fixed list of arbitrary sizes.

Luthaf commented:
int should not be smaller than 16-bits by the C standard, and not larger than 64 (maybe 128?), so we can just check for these sizes and be done =)

ezavod committed:
fix(cmake): restrict int size

Restrict allowed int sizes to 16, 32, 64 and 128 bit. This is necessary
because the offset calculation uses some bitwise operations to round to
the nearest int boundary. This only works if the size of int is a power
of two.

ezavod commented:
Added some checks to cmake, seems to do the trick. We need  at least [V3.3 because of `IN_LIST`](https://cmake.org/cmake/help/v3.3/release/3.3.html).
Just for reference: Minimum of 16bits for integers is described in 5.2.4.2.1 of [ISO/IEC9899:2017](https://web.archive.org/web/20181230041359if_/http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf).

Luthaf commented on CMakeLists.txt:
```diff
@@ -1,4 +1,6 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.3)
```
Why did you changed this? It looks like CheckTypeSize is supported with cmake 2.8.12 (https://cmake.org/cmake/help/v2.8.12/cmake.html#module:CheckTypeSize), which is the minimum required version here since it is the one shipped by some linux distributions (Debian/CentOS)

Luthaf replied:
Whoops, I did not see your other message on the justification ...

How about using [find](https://stackoverflow.com/a/23323790) instead?

Luthaf replied:
Or
```cmake
if (NOT "${INT_SIZE}" MATCHES "^(2|4|8|16)$")
    message(...)
endif()
```

ezavod replied:
Because i used `IN_LIST` in the added `if`block. Didn't know, that those older versions are still around. Could use `list(FIND ...)` but thought that this was kinda ugly.

Luthaf commented:
Thanks a lot for doing this!
